File: lc/py/020.py
# Single DP
class Solution:
    def climbStairs(self, n: int) -> int:
        prev = 1
        prev_prev = 1
        for i in range(1, n):
            curr = prev + prev_prev
            prev_prev = prev
            prev = curr
        return prev

# 1D DP
class Solution:
    def climbStairs(self, n: int) -> int:
        dp = [0] * n
        dp[0] = 1

        for i in range(1, n):
            dp[i] = dp[i-1]
            dp[i] += dp[i-2] if i-2 >= 0 else 1
        return dp[-1]

File: lc/c/020.c
// Single DP
int climbStairs(int n){
    int prev = 1, prev_prev = 1;
    for (int i = 1; i < n; ++i) {
        int curr = prev + prev_prev;
        prev_prev = prev;
        prev = curr;
    }
    return prev;
}

// 1D DP
int climbStairs(int n){
    int dp[n];
    memset(dp, 0, sizeof dp);
    dp[0] = 1;

    for (int i = 1; i < n; ++i) {
        dp[i] = dp[i-1];
        dp[i] += i-2 >= 0 ? dp[i-2] : 1;
    }
    return dp[n-1];
}
